File: src/ros2_msgs_enum_resolver.cpp
// Copyright 2024 Your Name
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "ros2_msgs_enum_resolver/ros2_msgs_enum_resolver.hpp"

#include <ament_index_cpp/get_package_share_directory.hpp>
#include <fstream>
#include <regex>
#include <sstream>
#include <string>

namespace ros2_msgs_enum_resolver
{

// ============================================================================
// Static members
// ============================================================================

std::mutex EnumResolver::s_mutex;
std::unordered_map<std::string, EnumResolver::ParsedConstants>
    EnumResolver::s_cache;
std::unordered_map<std::string, bool> EnumResolver::s_not_found;

// ============================================================================
// Public API
// ============================================================================

std::optional<std::vector<std::string>> EnumResolver::resolve(
    const std::string& interface_type,
    int64_t value,
    const std::string& field_type
)
{
  const auto* pc = getOrLoad(interface_type);
  if (!pc)
  {
    return std::nullopt;
  }

  // field_type provided -> search only within that declared type
  if (!field_type.empty())
  {
    // Check if by_type already has this field_type cached
    {
      std::lock_guard<std::mutex> lock(s_mutex);
      auto bt_it = pc->by_type.find(field_type);
      if (bt_it != pc->by_type.end())
      {
        const auto it = bt_it->second.find(value);
        if (it == bt_it->second.end())
        {
          return std::nullopt;
        }
        return it->second;
      }
    }

    // Need to build by_type for this field_type — reload section content
    TypeInfo info;
    if (!parseTypeString(interface_type, info))
    {
      return std::nullopt;
    }
    const std::string content = readFile(buildFilePath(info));
    if (content.empty())
    {
      return std::nullopt;
    }

    const auto sections = splitSections(content);
    std::size_t idx = 0;
    if (info.kind == "srv" && info.section == "Response")
    {
      idx = 1;
    }
    else if (info.kind == "action" && info.section == "Result")
    {
      idx = 1;
    }
    else if (info.kind == "action" && info.section == "Feedback")
    {
      idx = 2;
    }
    if (idx >= sections.size())
    {
      return std::nullopt;
    }

    auto by_type_map = parseSectionByType(sections[idx], field_type);

    {
      std::lock_guard<std::mutex> lock(s_mutex);
      // Non-const pointer needed to mutate cache entry
      auto& cached_pc = s_cache.at(interface_type);
      cached_pc.by_type[field_type] = by_type_map;
      const auto it = cached_pc.by_type[field_type].find(value);
      if (it == cached_pc.by_type[field_type].end())
      {
        return std::nullopt;
      }
      return it->second;
    }
  }

  // No field_type -> search all constants
  const auto it = pc->forward.find(value);
  if (it == pc->forward.end())
  {
    return std::nullopt;
  }
  return it->second;
}

std::optional<int64_t> EnumResolver::reverseResolve(
    const std::string& interface_type,
    const std::string& name
)
{
  const auto* pc = getOrLoad(interface_type);
  if (!pc)
  {
    return std::nullopt;
  }
  const auto it = pc->reverse.find(name);
  if (it == pc->reverse.end())
  {
    return std::nullopt;
  }
  return it->second;
}

std::optional<std::map<int64_t, std::vector<std::string>>>
EnumResolver::getConstants(const std::string& interface_type)
{
  const auto* pc = getOrLoad(interface_type);
  if (!pc || pc->forward.empty())
  {
    return std::nullopt;
  }
  return pc->forward;
}

std::optional<std::unordered_map<std::string, int64_t>>
EnumResolver::getConstantsReverse(const std::string& interface_type)
{
  const auto* pc = getOrLoad(interface_type);
  if (!pc || pc->reverse.empty())
  {
    return std::nullopt;
  }
  return pc->reverse;
}

std::optional<std::map<int64_t, std::vector<std::string>>>
EnumResolver::getConstantsForPrimitive(
    const std::string& interface_type,
    const std::string& field_type
)
{
  const auto* pc = getOrLoad(interface_type);
  if (!pc)
  {
    return std::nullopt;
  }

  // Check cache
  {
    std::lock_guard<std::mutex> lock(s_mutex);
    auto it = pc->by_type.find(field_type);
    if (it != pc->by_type.end())
    {
      if (it->second.empty())
      {
        return std::nullopt;
      }
      return it->second;
    }
  }

  // Build
  TypeInfo info;
  if (!parseTypeString(interface_type, info))
  {
    return std::nullopt;
  }
  const std::string content = readFile(buildFilePath(info));
  if (content.empty())
  {
    return std::nullopt;
  }

  const auto sections = splitSections(content);
  std::size_t idx = 0;
  if (info.kind == "srv" && info.section == "Response")
  {
    idx = 1;
  }
  else if (info.kind == "action" && info.section == "Result")
  {
    idx = 1;
  }
  else if (info.kind == "action" && info.section == "Feedback")
  {
    idx = 2;
  }
  if (idx >= sections.size())
  {
    return std::nullopt;
  }

  auto result = parseSectionByType(sections[idx], field_type);

  {
    std::lock_guard<std::mutex> lock(s_mutex);
    s_cache.at(interface_type).by_type[field_type] = result;
  }

  if (result.empty())
  {
    return std::nullopt;
  }
  return result;
}

bool EnumResolver::isEnumOnlyType(const std::string& interface_type)
{
  TypeInfo info;
  if (!parseTypeString(interface_type, info))
  {
    return false;
  }
  const std::string content = readFile(buildFilePath(info));
  if (content.empty())
  {
    return false;
  }

  const auto sections = splitSections(content);
  std::size_t idx = 0;
  if (info.kind == "srv" && info.section == "Response")
  {
    idx = 1;
  }
  else if (info.kind == "action" && info.section == "Result")
  {
    idx = 1;
  }
  else if (info.kind == "action" && info.section == "Feedback")
  {
    idx = 2;
  }
  if (idx >= sections.size())
  {
    return false;
  }

  return sectionHasOnlyConstants(sections[idx]);
}

bool EnumResolver::preload(const std::string& interface_type)
{
  return getOrLoad(interface_type) != nullptr;
}

void EnumResolver::clearCache()
{
  std::lock_guard<std::mutex> lock(s_mutex);
  s_cache.clear();
  s_not_found.clear();
}

bool EnumResolver::isAvailable(const std::string& interface_type)
{
  return getOrLoad(interface_type) != nullptr;
}

bool EnumResolver::isUnambiguous(const std::string& interface_type)
{
  const auto* pc = getOrLoad(interface_type);
  if (!pc)
  {
    return false;
  }
  for (const auto& kv : pc->forward)
  {
    if (kv.second.size() > 1)
    {
      return false;
    }
  }
  return true;
}

// ============================================================================
// Private helpers
// ============================================================================

bool EnumResolver::parseTypeString(const std::string& type, TypeInfo& out)
{
  std::vector<std::string> parts;
  {
    std::istringstream ss(type);
    std::string token;
    while (std::getline(ss, token, '/'))
    {
      if (!token.empty())
      {
        parts.push_back(token);
      }
    }
  }

  if (parts.size() < 3)
  {
    return false;
  }

  out.package = parts[0];
  out.kind = parts[1];
  out.name = parts[2];
  out.section = "";

  if (out.kind != "msg" && out.kind != "srv" && out.kind != "action")
  {
    return false;
  }

  if (parts.size() >= 4)
  {
    out.section = parts[3];
    if (out.kind == "srv")
    {
      if (out.section != "Request" && out.section != "Response")
      {
        return false;
      }
    }
    else if (out.kind == "action")
    {
      if (out.section != "Goal" && out.section != "Result" &&
          out.section != "Feedback")
      {
        return false;
      }
    }
    else
    {
      return false;
    }
  }

  return true;
}

std::string EnumResolver::buildFilePath(const TypeInfo& info)
{
  try
  {
    const std::string share =
        ament_index_cpp::get_package_share_directory(info.package);
    return share + "/" + info.kind + "/" + info.name + "." + info.kind;
  } catch (const std::exception&)
  {
    return {};
  }
}

std::string EnumResolver::readFile(const std::string& path)
{
  if (path.empty())
  {
    return {};
  }
  std::ifstream file(path);
  if (!file.is_open())
  {
    return {};
  }
  std::ostringstream buf;
  buf << file.rdbuf();
  return buf.str();
}

std::vector<std::string> EnumResolver::splitSections(const std::string& content)
{
  std::vector<std::string> sections;
  std::string current;
  std::istringstream stream(content);
  std::string line;

  while (std::getline(stream, line))
  {
    std::string trimmed = line;
    const auto first = trimmed.find_first_not_of(" \t\r\n");
    if (first == std::string::npos)
    {
      trimmed.clear();
    }
    else
    {
      trimmed = trimmed.substr(first);
      const auto last = trimmed.find_last_not_of(" \t\r\n");
      if (last != std::string::npos)
      {
        trimmed = trimmed.substr(0, last + 1);
      }
    }
    if (trimmed == "---")
    {
      sections.push_back(std::move(current));
      current.clear();
    }
    else
    {
      current += line;
      current += '\n';
    }
  }
  sections.push_back(std::move(current));
  return sections;
}

EnumResolver::ParsedConstants EnumResolver::parseSection(
    const std::string& section_content
)
{
  ParsedConstants result;

  // Matches:  <type>  <UPPER_NAME>  =  <value>
  static const std::regex kConstRegex(
      "^\\s*(?:bool|byte|char|float(?:32|64)|u?int(?:8|16|32|64))\\s+"
      "([A-Z][A-Z0-9_]*)\\s*=\\s*(-?(?:0[xX][0-9a-fA-F]+|\\d+))",
      std::regex::optimize
  );

  std::istringstream stream(section_content);
  std::string line;
  std::smatch match;

  while (std::getline(stream, line))
  {
    const auto cp = line.find('#');
    if (cp != std::string::npos)
    {
      line = line.substr(0, cp);
    }
    if (!std::regex_search(line, match, kConstRegex))
    {
      continue;
    }
    const std::string const_name = match[1].str();
    const int64_t const_value = std::stoll(match[2].str(), nullptr, 0);
    result.forward[const_value].push_back(const_name);
    result.reverse[const_name] = const_value;
  }
  return result;
}

std::map<int64_t, std::vector<std::string>> EnumResolver::parseSectionByType(
    const std::string& section_content,
    const std::string& field_type
)
{
  std::map<int64_t, std::vector<std::string>> result;

  // Matches:  <exact_type>  <UPPER_NAME>  =  <value>
  // We build the regex dynamically based on field_type.
  const std::string pattern =
      "^\\s*(" + field_type +
      ")\\s+"
      "([A-Z][A-Z0-9_]*)\\s*=\\s*(-?(?:0[xX][0-9a-fA-F]+|\\d+))";

  std::regex typed_re;
  try
  {
    typed_re = std::regex(pattern, std::regex::optimize);
  } catch (const std::regex_error&)
  {
    return result;
  }

  std::istringstream stream(section_content);
  std::string line;
  std::smatch match;

  while (std::getline(stream, line))
  {
    const auto cp = line.find('#');
    if (cp != std::string::npos)
    {
      line = line.substr(0, cp);
    }
    if (!std::regex_search(line, match, typed_re))
    {
      continue;
    }
    const std::string const_name = match[2].str();
    const int64_t const_value = std::stoll(match[3].str(), nullptr, 0);
    result[const_value].push_back(const_name);
  }
  return result;
}

bool EnumResolver::sectionHasOnlyConstants(const std::string& section_content)
{
  // A regular (non-constant) field line looks like:
  //   <type> <lowercase_name>
  // without an '=' sign.
  static const std::regex kFieldRegex(
      "^\\s*(?:bool|byte|char|string|wstring|float(?:32|64)|u?int(?:8|16|32|64)"
      "|[a-zA-Z][a-zA-Z0-9_]*/[a-zA-Z][a-zA-Z0-9_/]*)"
      "\\s+[a-z][a-zA-Z0-9_]*(\\[.*?\\])?(\\s+\\S+)?\\s*$",
      std::regex::optimize
  );

  bool has_any = false;
  std::istringstream stream(section_content);
  std::string line;

  while (std::getline(stream, line))
  {
    const auto cp = line.find('#');
    if (cp != std::string::npos)
    {
      line = line.substr(0, cp);
    }
    const auto first = line.find_first_not_of(" \t\r\n");
    if (first == std::string::npos)
    {
      continue;
    }
    has_any = true;
    // constant lines always contain '='
    if (line.find('=') != std::string::npos)
    {
      continue;
    }
    // line has no '=' -> check if it looks like a field
    if (std::regex_search(line, kFieldRegex))
    {
      return false;
    }
  }
  return has_any;
}

std::optional<EnumResolver::ParsedConstants> EnumResolver::loadType(
    const std::string& interface_type
)
{
  TypeInfo info;
  if (!parseTypeString(interface_type, info))
  {
    return std::nullopt;
  }

  const std::string content = readFile(buildFilePath(info));
  if (content.empty())
  {
    return std::nullopt;
  }

  const auto sections = splitSections(content);

  std::size_t idx = 0;
  if (info.kind == "srv" && info.section == "Response")
  {
    idx = 1;
  }
  else if (info.kind == "action" && info.section == "Result")
  {
    idx = 1;
  }
  else if (info.kind == "action" && info.section == "Feedback")
  {
    idx = 2;
  }

  if (idx >= sections.size())
  {
    return std::nullopt;
  }
  return parseSection(sections[idx]);
}

const EnumResolver::ParsedConstants* EnumResolver::getOrLoad(
    const std::string& interface_type
)
{
  // Fast path
  {
    std::lock_guard<std::mutex> lock(s_mutex);
    const auto it = s_cache.find(interface_type);
    if (it != s_cache.end())
    {
      return &it->second;
    }
    if (s_not_found.count(interface_type))
    {
      return nullptr;
    }
  }

  // Slow path - I/O outside lock
  auto parsed = loadType(interface_type);

  // Insert under lock with double-check
  {
    std::lock_guard<std::mutex> lock(s_mutex);
    const auto it = s_cache.find(interface_type);
    if (it != s_cache.end())
    {
      return &it->second;
    }
    if (parsed)
    {
      auto& ref = s_cache[interface_type];
      ref = std::move(*parsed);
      return &ref;
    }
    s_not_found[interface_type] = true;
    return nullptr;
  }
}

}  // namespace ros2_msgs_enum_resolver

File: ros2_msgs_enum_resolver/ros2_msgs_enum_resolver.py
# Copyright 2024 Your Name
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""ros2_msgs_enum_resolver - Python EnumResolver.

Supported type strings::

    pkg/msg/TypeName
    pkg/srv/TypeName[/Request|/Response]          default: Request
    pkg/action/TypeName[/Goal|/Result|/Feedback]  default: Goal

Example
-------
::

    from ros2_msgs_enum_resolver import EnumResolver

    # Enum-only message type (field type IS the enum)
    names = EnumResolver.resolve('ura_msgs/msg/ServerConnectionEventEnum', 1)

    # Inline constants (NavSatStatus pattern)
    names = EnumResolver.resolve('sensor_msgs/msg/NavSatStatus', 0, 'int8')

"""

from __future__ import annotations

from pathlib import Path
import re
import threading
from typing import Dict, List, Optional

from ament_index_python.packages import (
    get_package_share_directory,
    PackageNotFoundError,
)

_VALID_KINDS = {'msg', 'srv', 'action'}
_SRV_SECTIONS = {'Request': 0, 'Response': 1}
_ACTION_SECTIONS = {'Goal': 0, 'Result': 1, 'Feedback': 2}

# Matches any constant (all types)
_CONST_RE = re.compile(
    r'^\s*(?:bool|byte|char|float(?:32|64)|u?int(?:8|16|32|64))\s+'
    r'([A-Z][A-Z0-9_]*)\s*=\s*(-?(?:0[xX][0-9a-fA-F]+|\d+))'
)

# Matches a constant and captures its declared type
_TYPED_CONST_RE = re.compile(
    r'^\s*(bool|byte|char|float(?:32|64)|u?int(?:8|16|32|64))\s+'
    r'([A-Z][A-Z0-9_]*)\s*=\s*(-?(?:0[xX][0-9a-fA-F]+|\d+))'
)

# Matches a regular field (lowercase name, no '=' sign)
_FIELD_RE = re.compile(
    r'^\s*(?:bool|byte|char|string|wstring|float(?:32|64)|u?int(?:8|16|32|64)'
    r'|[a-zA-Z][a-zA-Z0-9_]*/[a-zA-Z][a-zA-Z0-9_/]*)'
    r'\s+[a-z][a-zA-Z0-9_]*(\[.*?\])?\s*$'
)


class _TypeInfo:
    __slots__ = ('package', 'kind', 'name', 'section')

    def __init__(self, package: str, kind: str, name: str, section: str) -> None:
        self.package = package
        self.kind = kind
        self.name = name
        self.section = section


class _ParsedConstants:
    __slots__ = ('forward', 'reverse', 'by_type')

    def __init__(self) -> None:
        self.forward: Dict[int, List[str]] = {}
        self.reverse: Dict[str, int] = {}
        # primitive_type -> {value: [name, ...]}
        self.by_type: Dict[str, Dict[int, List[str]]] = {}


class EnumResolver:
    """Static class - use class methods directly, no instantiation needed.

    Thread-safe lazy cache backed by threading.Lock.
    """

    _lock: threading.Lock = threading.Lock()
    _cache: Dict[str, _ParsedConstants] = {}
    _not_found: set = set()

    @classmethod
    def resolve(
        cls,
        interface_type: str,
        value: int,
        field_type: str = '',
    ) -> Optional[List[str]]:
        """Resolve value to list of names, or None if not found.

        When field_type is empty, searches all constants in the interface.
        When field_type is a primitive type name such as 'int8', only
        constants declared with that exact type are searched.
        """
        pc = cls._get_or_load(interface_type)
        if pc is None:
            return None

        if field_type:
            by_type_map = cls._get_by_type(interface_type, pc, field_type)
            if by_type_map is None:
                return None
            return by_type_map.get(value)

        return pc.forward.get(value)

    @classmethod
    def reverse_resolve(cls, interface_type: str, name: str) -> Optional[int]:
        """Resolve name to numeric value, or None if not found."""
        pc = cls._get_or_load(interface_type)
        if pc is None:
            return None
        return pc.reverse.get(name)

    @classmethod
    def get_constants(cls, interface_type: str) -> Optional[Dict[int, List[str]]]:
        """Return all constants sorted by value, or None if not found."""
        pc = cls._get_or_load(interface_type)
        if pc is None or not pc.forward:
            return None
        return dict(sorted(pc.forward.items()))

    @classmethod
    def get_constants_reverse(cls, interface_type: str) -> Optional[Dict[str, int]]:
        """Return all constants as name->value dict, or None if not found."""
        pc = cls._get_or_load(interface_type)
        if pc is None or not pc.reverse:
            return None
        return dict(pc.reverse)

    @classmethod
    def get_constants_for_primitive(
        cls,
        interface_type: str,
        field_type: str,
    ) -> Optional[Dict[int, List[str]]]:
        """Return constants whose declared type matches field_type exactly."""
        pc = cls._get_or_load(interface_type)
        if pc is None:
            return None
        result = cls._get_by_type(interface_type, pc, field_type)
        if not result:
            return None
        return dict(sorted(result.items()))

    @classmethod
    def is_enum_only_type(cls, interface_type: str) -> bool:
        """Check whether the interface has only constants and no regular fields."""
        info = cls._parse_type_string(interface_type)
        if info is None:
            return False
        path = cls._build_file_path(info)
        if path is None:
            return False
        sections = cls._split_sections(path.read_text(encoding='utf-8'))
        if info.kind == 'srv':
            idx = _SRV_SECTIONS.get(info.section, 0)
        elif info.kind == 'action':
            idx = _ACTION_SECTIONS.get(info.section, 0)
        else:
            idx = 0
        if idx >= len(sections):
            return False
        return cls._section_has_only_constants(sections[idx])

    @classmethod
    def preload(cls, interface_type: str) -> bool:
        """Pre-warm the cache for the given type. Return True on success."""
        return cls._get_or_load(interface_type) is not None

    @classmethod
    def clear_cache(cls) -> None:
        """Flush the entire in-memory cache."""
        with cls._lock:
            cls._cache.clear()
            cls._not_found.clear()

    @classmethod
    def is_available(cls, interface_type: str) -> bool:
        """Check whether the type resolves to a loadable interface file."""
        return cls._get_or_load(interface_type) is not None

    @classmethod
    def is_unambiguous(cls, interface_type: str) -> bool:
        """Check whether every value maps to exactly one name (no aliases)."""
        pc = cls._get_or_load(interface_type)
        if pc is None:
            return False
        return all(len(names) == 1 for names in pc.forward.values())

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    @classmethod
    def _get_or_load(cls, interface_type: str) -> Optional[_ParsedConstants]:
        with cls._lock:
            if interface_type in cls._cache:
                return cls._cache[interface_type]
            if interface_type in cls._not_found:
                return None
        parsed = cls._load_type(interface_type)
        with cls._lock:
            if interface_type in cls._cache:
                return cls._cache[interface_type]
            if parsed is not None:
                cls._cache[interface_type] = parsed
                return parsed
            cls._not_found.add(interface_type)
            return None

    @classmethod
    def _get_by_type(
        cls,
        interface_type: str,
        pc: _ParsedConstants,
        field_type: str,
    ) -> Optional[Dict[int, List[str]]]:
        """Return (and cache) constants filtered by declared primitive type."""
        with cls._lock:
            if field_type in pc.by_type:
                return pc.by_type[field_type] or None

        info = cls._parse_type_string(interface_type)
        if info is None:
            return None
        path = cls._build_file_path(info)
        if path is None:
            return None

        sections = cls._split_sections(path.read_text(encoding='utf-8'))
        if info.kind == 'srv':
            idx = _SRV_SECTIONS.get(info.section, 0)
        elif info.kind == 'action':
            idx = _ACTION_SECTIONS.get(info.section, 0)
        else:
            idx = 0
        if idx >= len(sections):
            return None

        result = cls._parse_section_by_type(sections[idx], field_type)
        with cls._lock:
            pc.by_type[field_type] = result
        return result or None

    @staticmethod
    def _parse_type_string(s: str) -> Optional[_TypeInfo]:
        parts = [p for p in s.split('/') if p]
        if len(parts) < 3:
            return None
        package, kind, name = parts[0], parts[1], parts[2]
        section = ''
        if kind not in _VALID_KINDS:
            return None
        if len(parts) >= 4:
            section = parts[3]
            if kind == 'srv' and section not in _SRV_SECTIONS:
                return None
            if kind == 'action' and section not in _ACTION_SECTIONS:
                return None
            if kind == 'msg':
                return None
        return _TypeInfo(package, kind, name, section)

    @staticmethod
    def _build_file_path(info: _TypeInfo) -> Optional[Path]:
        try:
            share = get_package_share_directory(info.package)
        except PackageNotFoundError:
            return None
        p = Path(share) / info.kind / f'{info.name}.{info.kind}'
        return p if p.exists() else None

    @staticmethod
    def _split_sections(content: str) -> List[str]:
        sections: List[str] = []
        current: List[str] = []
        for line in content.splitlines():
            if line.strip() == '---':
                sections.append('\n'.join(current))
                current = []
            else:
                current.append(line)
        sections.append('\n'.join(current))
        return sections

    @staticmethod
    def _parse_section(section_content: str) -> _ParsedConstants:
        pc = _ParsedConstants()
        for line in section_content.splitlines():
            comment = line.find('#')
            if comment != -1:
                line = line[:comment]
            m = _CONST_RE.match(line)
            if not m:
                continue
            const_name = m.group(1)
            const_value = int(m.group(2), 0)
            pc.forward.setdefault(const_value, []).append(const_name)
            pc.reverse[const_name] = const_value
        return pc

    @staticmethod
    def _parse_section_by_type(
        section_content: str,
        field_type: str,
    ) -> Dict[int, List[str]]:
        """Parse constants whose declared type matches field_type exactly."""
        result: Dict[int, List[str]] = {}
        for line in section_content.splitlines():
            comment = line.find('#')
            if comment != -1:
                line = line[:comment]
            m = _TYPED_CONST_RE.match(line)
            if not m:
                continue
            if m.group(1) != field_type:
                continue
            const_name = m.group(2)
            const_value = int(m.group(3), 0)
            result.setdefault(const_value, []).append(const_name)
        return result

    @staticmethod
    def _section_has_only_constants(section_content: str) -> bool:
        """Return True if every meaningful line is a constant declaration."""
        has_any = False
        for line in section_content.splitlines():
            stripped = line.strip()
            if not stripped or stripped.startswith('#'):
                continue
            # strip inline comment
            no_comment = stripped.split('#')[0].strip()
            if not no_comment:
                continue
            has_any = True
            # constant lines always contain '='
            if '=' in no_comment:
                continue
            # regular field - check it matches field pattern
            if _FIELD_RE.match(line):
                return False
        return has_any

    @classmethod
    def _load_type(cls, interface_type: str) -> Optional[_ParsedConstants]:
        info = cls._parse_type_string(interface_type)
        if info is None:
            return None
        path = cls._build_file_path(info)
        if path is None:
            return None
        sections = cls._split_sections(path.read_text(encoding='utf-8'))
        idx = 0
        if info.kind == 'srv' and info.section == 'Response':
            idx = 1
        elif info.kind == 'action' and info.section == 'Result':
            idx = 1
        elif info.kind == 'action' and info.section == 'Feedback':
            idx = 2
        if idx >= len(sections):
            return None
        return cls._parse_section(sections[idx])